### ISSUE TITLE
[Major Bug/Skill Issue] Incorrect Benchmark Methodology - `time` Command Misrepresents Java/Julia/etc Performance

### ISSUE THREAD

**Issue by wiseaidev** [bddicken/languages#149]
**Description:**

The current benchmarking approach using the `time` command is inaccurate for Java/Julia/etc performance evaluation. It includes JIT compilation and warm-up time, leading to misleading results.

**Recommendation:**

Instead of using `time`, implement a custom benchmark that measures the internal execution time of the program for each programming language. This approach avoids external overhead and provides more precise results by focusing solely on the core execution, excluding JIT and warm-up phases. This will ensure accurate and consistent performance measurements. Take the following Java bench example excluding JIT:

```java
public class Fibonacci {
    private static int fibonacci(int n) {
        if (n == 0) {
            return 0;
        }
        if (n == 1) {
            return 1;
        }
        return fibonacci(n - 1) + fibonacci(n - 2);
    }

    public static void main(String[] args) {
        final int N = Integer.parseInt(args[0]);
        double total = 0.0;

        // 👇 JIT warm-up phase to ensure JVM optimizations are fully applied.
        for (int i = 0; i < 1_000_000; i++) {
            fibonacci(10); // 👈 Arbitrary Fibonacci call to trigger JIT optimizations.
        }

        final int benchmarkIterations = 20; // 👈 Total runs

        for (int count = 1; count <= benchmarkIterations; count++) {
            long startTime = System.nanoTime();
            int result = fibonacci(N);
            long endTime = System.nanoTime();
            double elapsed = (endTime - startTime) / 1_000_000_000.0;
            total += elapsed;
        }

        double avgTime = total / benchmarkIterations; // 👈 Average after warm-up.
        System.out.printf("🚀 Average time taken: %.6f s%n", avgTime);
    }
}

// javac Fibonacci.java && java Fibonacci 40
```

Besides, I suggest exploring different JVM command line options to further optimize performance during benchmarking (e.g. Garbage Collector Tuning: https://www.oracle.com/technical-resources/articles/java/g1gc.html).

This will provide more valid and reproducible results.

**Comment by jrosell** [comment]
Yes. Related https://github.com/bddicken/languages/issues/98

**Comment by bddicken** [comment]
I understand there's a difference between timing full process execution times and just code snippet times. For the time being, I intend to keep doing full process execution time. Yes, this includes overheads such as interpreter startup and jit compilation times. Other benchmarks may choose to ignore this, but in my case that is part of the calculation. If you choose to use an interpreted languages, great! But that comes at a cost of additional over head to get to the execution of your actual code.

**Comment by jrosell** [comment]
Then, why not including compilation time too?

**Comment by harbulot** [comment]
@bddicken You're right, both approaches can make sense: full process execution time or just the loop within its application.

> If you choose to use an interpreted languages, great! But that comes at a cost of additional over head to get to the execution of your actual code.

The problem here is that you seem to be treating Java as an "interpreted" language: it isn't (at least, it's far more subtle than that).
The whole purpose of languages with a JIT compiler within their runtime environment is to benefit from optimisations that were not anticipated at compile time (or load time if the code comes from source directly), at the expense of an initial delay.

Java's HotSpot has various options to tweak the performance, [in particular `-server`](https://www.oracle.com/java/technologies/whitepaper.html): "*the Server VM has been specially tuned to maximize peak operating speed. It is intended for executing long-running server applications, which need the fastest possible operating speed more than a fast start-up time or smaller runtime memory footprint.*"

In short, these systems have been designed knowing that the application would take a performance hit at startup, but with the knowledge (or the hope) that it would bring improvements in the long run.

There have been significant changes in performance across JVM versions for example. Even for the same major and minor version, you might have different default settings depending on the vendor.
I would guess you can have similar issues for other languages too.

It's very difficult to benchmark languages against each other. Ideally, you'd want to pick the best version for each, within its target conditions, but there can be plenty of variations. Even in C, [`-O3` can be slower than `-O2` in some cases](https://stackoverflow.com/questions/28875325/gcc-optimization-flag-o3-makes-code-slower-than-o2).

**Comment by mkitti** [comment]
In the Julia case, there's an ahead-of-time compilation mode that could be used:
https://github.com/JuliaLang/julia/blob/master/contrib/juliac.jl

**Comment by bddicken** [comment]
Thanks for the discussion here. Closing this out for a few reasons:

- (A) We're now using hyperfine instead of time.
- (B) For now we are still including startup time, but there been discussion around how to account for this in the timings. For example, see: https://github.com/bddicken/languages/issues/220

**Comment by stevekuznetsov** [comment]
You could of course also provide enough input data such that your benchmarks operate over minutes or hours, or you could measure output throughput, _etc_. The choice of inputs that leads to millisecond start-run-stop times furthermore increases the impact of interrupts, system daemons waking up, whatever, just as it increases the impact of interpreter startup in the overall number being measured.
